outbounds/proxyless/proxyless.go

```diff
@@ -0,0 +1,104 @@
+	"github.com/getlantern/radiance/transport/proxyless"
+	"github.com/sagernet/sing-box/adapter"
+	"github.com/sagernet/sing-box/adapter/outbound"
+	"github.com/sagernet/sing-box/common/dialer"
+	"github.com/sagernet/sing-box/log"
+	"github.com/sagernet/sing-box/option"
+	"github.com/sagernet/sing/common/logger"
+	"github.com/sagernet/sing/common/metadata"
+	"github.com/sagernet/sing/common/network"
+)
+
+var glog = golog.LoggerFor("proxyless")
+
+// ProxylessOutbound implements a proxyless outbound
+type ProxylessOutbound struct {
+	outbound.Adapter
+	logger logger.ContextLogger
+	dialer otransport.StreamDialer
+}
+
+// NewProxylessOutbound creates a proxyless outbond that uses the proxyless transport
+// for dialing
+func NewProxylessOutbound(ctx context.Context, router adapter.Router, logger log.ContextLogger, tag string, options option.DialerOptions) (adapter.Outbound, error) {
+	glog.Debug("creating outbound dialer")
+	outboundDialer, err := dialer.New(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+
+	glog.Debug("getting config")
+	ch := config.NewConfigHandler(time.Minute * 10)
+	configCtx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
+	conf, err := ch.GetConfig(configCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	outSD := &sboxSD{
+		outSD:  outboundDialer,
+		logger: logger,
+	}
+	dialer, err := proxyless.NewStreamDialer(outSD, conf[0])
+	if err != nil {
+		return nil, err
+	}
+
+	return &ProxylessOutbound{
+		Adapter: outbound.NewAdapterWithDialerOptions("proxyless", tag, []string{network.NetworkTCP}, options),
+		logger:  logger,
+		dialer:  dialer,
+	}, nil
+}
+
+// DialContext extracts the metadata domain, add the destination to the context
+// and use the proxyless dialer for sending the request
+func (o *ProxylessOutbound) DialContext(ctx context.Context, network string, destination metadata.Socksaddr) (net.Conn, error) {
+	ctx, metadata := adapter.ExtendContext(ctx)
+	metadata.Outbound = o.Tag()
+	metadata.Destination = destination
+	ctx = context.WithValue(ctx, proxyless.DomainContextKey, metadata.Domain)
+	o.logger.InfoContext(ctx, "received proxyless request to %q (%q) domain", metadata.Domain, destination.String())
+	conn, err := o.dialer.DialStream(ctx, destination.String())
+	if err != nil {
+		o.logger.ErrorContext(ctx, "failed to dial to %q: %w", metadata.Domain, err)
+	}
+	return conn, err
+}
+
+// ListenPacket isn't implemented
+func (o *ProxylessOutbound) ListenPacket(ctx context.Context, destination metadata.Socksaddr) (net.PacketConn, error) {
+	return nil, os.ErrInvalid
+}
+
+// wrapper around sing-box's network.Dialer to implement streamDialer interface to pass to a
+// stream dialer as innerSD
+type sboxSD struct {
+	outSD  network.Dialer
+	logger log.ContextLogger
+}
+
+func (s *sboxSD) DialStream(ctx context.Context, addr string) (otransport.StreamConn, error) {
+	s.logger.InfoContext(ctx, "proxyless sboxSD dialing ", addr)
+	destination := metadata.ParseSocksaddr(addr)
```

Comment:
Just curious -- why is the input a SOCKS address?

Reply:
The `network.Dialer` interface from `github.com/sagernet/sing/common/network` expects a SocksAddr (in the same way as the outbound does) but I don't know if there's an sing-box special reason